src/os/bluestore/BlueFS.cc

```diff
@@ -606,6 +613,140 @@ void BlueFS::_stop_alloc()
   }
 }
 
+int BlueFS::read(uint8_t ndev, uint64_t off, uint64_t len,
+		 ceph::buffer::list *pbl, IOContext *ioc, bool buffered)
+{
+  int r;
+  bufferlist bl;
+  r = bdev[ndev]->read(off, len, &bl, ioc, buffered);
+  if (r != 0) {
+    return r;
+  }
+  uint64_t block_size = bdev[ndev]->get_block_size();
+  if (inject_read_zeros) {
+    if (len >= block_size * 2) {
+      //use beginning, replace 8K in the middle with zeros, use tail
+      bufferlist temp;
+      bl.splice(0, len / 2 - block_size, &temp);
+      temp.append(buffer::create(block_size * 2, 0));
+      bl.splice(block_size * 2, len / 2 - block_size, &temp);
+      bl = temp;
+      inject_read_zeros--;
+    }
+  }
+  //make a check if there is a block with all 0
+  uint64_t to_check_len = len;
+  uint64_t skip = p2nphase(len, block_size);
+  if (skip > to_check_len) {
+    return r;
+  }
+  auto it = bl.begin(skip);
+  to_check_len -= skip;
+  bool all_zeros = false;
+  while (all_zeros == false && to_check_len > block_size) {
+    // checking 0s step
+    unsigned block_left = block_size;
+    unsigned avail;
+    const char* data;
+    all_zeros = true;
+    while (all_zeros && block_left > 0) {
+      avail = it.get_ptr_and_advance(block_left, &data);
+      for (unsigned i = 0 ; i < avail; i++) {
+	if (data[i] != 0) {
+	  all_zeros = false;
+	  break;
+	}
+      }
+      block_left -= avail;
+    }
+    // skipping step
+    while (block_left > 0) {
+      avail = it.get_ptr_and_advance(block_left, &data);
+      block_left -= avail;
+    }
+    to_check_len -= block_size;
+  }
+  if (all_zeros) {
+    logger->inc(l_bluefs_read_zeros_candidate, 1);
+    bufferlist bl_reread;
+    r = bdev[ndev]->read(off, len, &bl_reread, ioc, buffered);
+    if (r != 0) {
+      return r;
+    }
+    // check if both read gave the same
+    // it is expected to be infrequent, so we sacrifice performance for simplicity
+    char* bl_data = bl.c_str();
+    char* bl_reread_data = bl_reread.c_str();
+    if (memcmp(bl_data, bl_reread_data, len) != 0) {
+	// report problems to log, but continue, maybe it will be good now...
+	derr << __func__ << " initial read of " << int(ndev)
```

Comment:
suggest to introduce another perf counter, e.g. bluefs_read_zeros_errors and increment it at this point as well. Sometimes it's easier to analyze perf counters....

Reply:
Good idea. Added.